playground/assets/__tests__/assets.spec.ts

```diff
@@ -24,19 +24,62 @@ const assetMatch = isBuild
 
 const iconMatch = `/foo/icon.png`
 
+const fetchPath = (p: string) => {
+  return fetch(path.posix.join(viteTestUrl, p), {
+    headers: { Accept: 'text/html,*/*' },
+  })
+}
+
 test('should have no 404s', () => {
   browserLogs.forEach((msg) => {
     expect(msg).not.toMatch('404')
   })
 })
 
 test('should get a 404 when using incorrect case', async () => {
-  expect((await fetch(path.posix.join(viteTestUrl, 'icon.png'))).status).toBe(
-    200,
-  )
-  expect((await fetch(path.posix.join(viteTestUrl, 'ICON.png'))).status).toBe(
-    404,
-  )
+  expect((await fetchPath('icon.png')).status).toBe(200)
+  // won't be wrote to index.html because the url includes `.`
+  expect((await fetchPath('ICON.png')).status).toBe(404)
+
+  expect((await fetchPath('bar')).status).toBe(200)
+  // fallback to index.html
+  const incorrectBarFetch = await fetchPath('BAR')
+  expect(incorrectBarFetch.status).toBe(200)
+  expect(incorrectBarFetch.headers.get('Content-Type')).toBe('text/html')
+})
+
+test('html resolve behavior', async () => {
+  const [
+    nestedIndexHtml,
+    nested,
+    nestedSlash,
+
+    nonNestedHtml,
+    nonNested,
+    nonNestedSlash,
+  ] = await Promise.all([
+    fetchPath('nested/index.html'), // -> nested/index.html
+    fetchPath('nested'), // -> index.html
+    fetchPath('nested/'), // -> nested/index.html
+
+    fetchPath('non-nested.html'), // -> non-nested.html
+    fetchPath('non-nested'), // -> index.html
+    fetchPath('non-nested/'), // -> index.html
```

Comment:
With dev server, these tests passes without this PR.
With preview server, these tests resolved to

- `nested/index.html` -> `nested/index.html`
- `nested` -> `nested/index.html` (**changed**, #6596)
- `nested/` -> `nested/index.html`
- `non-nested.html` -> `non-nested.html`
- `non-nested` -> 404 (**changed**, #10925)
- `non-nested/` -> 404 (**changed**, #10925)